### PR TITLE
Change the way singular matrices are detected in MathUtils::InvertMatrix

### PR DESCRIPTION
**Pull request by pablobecker**
Currently, MathUtils is attempting to detect singular matrix by simply looking at the determinant. If it is lower than approx 1e-16, it reports singular matrix and throws and error.
THIS IS INCORRECT. If a matrix has all its coeffs , for example, in the order 1e-10, the determinant will be of order (1e-10)^n , where n is the size of the matrix. 

The correct way to do it is with the condition number.
I have removed the check with the determinant and added the condition number. It only implies a cheap operation, computing the norm of the two matrices.

NOTE: I went for the norm  sqrt (sum I(i,j) ) due to simplicity and cost. If mathematically it is better to use another norm, please provide a link with the best norm / change it yourself.

### CODE REVIEW

File: kratos/utilities/math_utils.h [hunk 1]
```diff
@@ -279,20 +279,14 @@ class MathUtils
     static inline BoundedMatrix<TDataType, TDim, TDim> InvertMatrix(
             const BoundedMatrix<TDataType, TDim, TDim>& InputMatrix,
             TDataType& InputMatrixDet,
-            const TDataType Tolerance = ZeroTolerance
+            const TDataType MaxCondNumber = (1.0/std::numeric_limits<double>::epsilon()) * 1.0e-4 // we want at least 4 significant digits.
```
Comment (loumalouomega) [inline kratos/utilities/math_utils.h]: Use max instead https://en.cppreference.com/w/cpp/types/numeric_limits/max

**Review by loumalouomega — CHANGES_REQUESTED** [review]

File: kratos/utilities/math_utils.h [hunk 2]
```diff
@@ -10,7 +10,7 @@
 //  Main authors:    Pooyan Dadvand
 //                   Riccardo Rossi
 //
-//  Collaborator:    Vicente Mataix Ferrandiz
+//  Collaborators:    Vicente Mataix Ferrandiz , pbecker
```
Comment (philbucher) [inline kratos/utilities/math_utils.h]: please use full names

File: kratos/utilities/math_utils.h [hunk 3]
```diff
@@ -357,6 +351,28 @@ class MathUtils
...
 
+        //Find the condition number to define is inverse is OK
+        double input_matrix_norm = 0.0;
+        double inverted_matrix_norm = 0.0;
+        for(unsigned int i = 0; i < TDim; i++)
+        {
+            for(unsigned int j = 0; j < TDim; j++)
+            {
+                input_matrix_norm += std::pow(InputMatrix(i,j),2);
+                inverted_matrix_norm += std::pow(InvertedMatrix(i,j),2);
```
Comment (philbucher) [inline kratos/utilities/math_utils.h]: Is this method not yet implemented somewhere else?

Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: Yes, but for sparse matrices

File: kratos/utilities/math_utils.h (hunk 1 above)
Comment (pablobecker) [reply kratos/utilities/math_utils.h]: no, the max is 1e308, I need the highest value that can be compared against 1. The inverse of epsilon, around 1e16.

**Review by loumalouomega — DISMISSED** [review]
Approving then

**Comment by philbucher** [comment]
tests are running @pablobecker ?

**Comment by pablobecker** [comment]
@philbucher Let me search which other apps are using it , to confirmo 100%

File: kratos/utilities/math_utils.h (hunk 3 above)
Comment (philbucher) [reply kratos/utilities/math_utils.h]: wow :)
The entire Kratos does not have an implementation for the L2-Norm? :D

Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: This way the double loop is done only once

**Comment by philbucher** [comment]
Tests in Travis are failing

**Comment by loumalouomega** [comment]
58.6482 is not so high, probably the behavior change is the reason

**Comment by RiccardoRossi** [comment]
for a matrix it is called "norm_frobenius"

      double norm = norm_frobenius(A)

it is in ublas

**Comment by loumalouomega** [comment]
But you are computing two norms, this is slightly more efficient?

**Comment by pablobecker** [comment]
Thanks for the changes @loumalouomega . It should work ok now.

**Review by loumalouomega — DISMISSED** [review]

**Comment by loumalouomega** [comment]
> Thanks for the changes @loumalouomega . It should work ok now.

imagen

**Comment by pablobecker** [comment]
Once tests are passed, we can remove the [Behaviour Change] tag

**Comment by philbucher** [comment]
> Once tests are passed, we can remove the [Behaviour Change] tag

@pablobecker the behavior does change, so this tag should not be removed (previously you used the det to check, now you use the cond-number)
One of the ideas of the tags is that for a release all PRs with certain Tags can be grouped

**Review by loumalouomega — APPROVED** [review]

**Comment by loumalouomega** [comment]
@pablobecker I hope this is the last correction

**Comment by pablobecker** [comment]
I was just looking at the same. I found that the spr_error_process.cpp had a negative tolerance. I'm not sure if I agree that negative tolerance implies turning the check off, but if it is being used this way, then it's probably better to stick to the currentbehavior..

**Comment by loumalouomega** [comment]
> I was just looking at the same. I found that the spr_error_process.cpp had a negative tolerance. I'm not sure if I agree that negative tolerance implies turning the check off, but if it is being used this way, then it's probably better to stick to the currentbehavior..

Indeed

**Comment by loumalouomega** [comment]
Now it works

**Pull request merged**